lib/transformer.cc

```diff
@@ -0,0 +1,176 @@
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://gnu.org/licenses/gpl-2.0.txt>
+
+#include <assert.h>
+
+#include "transformer.h"
+
+namespace rgb_matrix {
+
+RotateCanvas::RotateCanvas(int angle) : delegatee_(NULL), angle_(angle) {
+  // Only allow angles in 90° steps
+  assert(angle % 90 == 0);
+}
+
+void RotateCanvas::SetDelegatee(Canvas* delegatee) {
+  assert(delegatee != NULL);
+
+  pivot_x_ = (delegatee->width() - 1) / 2;
+  pivot_y_ = (delegatee->height() - 1) / 2;
+  delegatee_ = delegatee;
+}
+
+void RotateCanvas::SetPixel(int x, int y, uint8_t red, uint8_t green, uint8_t blue) {
+  assert(delegatee_ != NULL);
+
+  const float s = (angle_ == 90 ? 1 : (angle_ == 270 ? -1 : 0));
+  const float c = (angle_ == 180 ? -1 : (angle_ == 0 ? 1 : 0));
+
+  // Offset needed cause of little precision errors on 180° and 270°
+  const int offset_x = (angle_ == 90 || angle_ == 180 ? 1 : 0);
+  const int offset_y = (angle_ == 180 || angle_ == 270 ? 1 : 0);
+
+  // translate point to origin
+  x -= pivot_x_;
+  y -= pivot_y_;
+
+  float rot_x = x * c - y * s;
+  float rot_y = x * s + y * c;
+
+  // translate back
+  x = rot_x + pivot_x_ + offset_x;
+  y = rot_y + pivot_y_ + offset_y;
+
+  delegatee_->SetPixel(x, y, red, green, blue);
+}
+
+int RotateCanvas::width() const { 
+  assert(delegatee_ != NULL);
+  return (angle_ % 180 == 0) ? delegatee_->width() : delegatee_->height();
+}
+
+int RotateCanvas::height() const { 
+  assert(delegatee_ != NULL); 
+  return (angle_ % 180 == 0) ? delegatee_->height() : delegatee_->width();
+}
+
+void RotateCanvas::Clear() { 
+  assert(delegatee_ != NULL); 
+  delegatee_->Clear(); 
+}
+
+void RotateCanvas::Fill(uint8_t red, uint8_t green, uint8_t blue) {
+  assert(delegatee_ != NULL);
+  delegatee_->Fill(red, green, blue);
+}
+
+void RotateCanvas::SetAngle(int angle) {
+  assert(angle % 90 == 0);
+  angle_ = angle;
+}
+
+int RotateCanvas::angle() {
+  return angle_;
+}
+
+RotateCanvasTransformer::RotateCanvasTransformer(int angle) : angle_(angle) {
+  assert(angle % 90 == 0);
+  canvas_ = new RotateCanvas(angle);
+}
+
+Canvas *RotateCanvasTransformer::Transform(Canvas *output) {
+  canvas_->SetDelegatee(output);
+  return canvas_;
+}
+
+void RotateCanvasTransformer::SetAngle(int angle) {
+  assert(angle % 90 == 0);
+  canvas_->SetAngle(angle);
+  angle_ = angle;
+}
+
+int RotateCanvasTransformer::angle() {
+  return angle_;
+}
+
+void LinkedTransformerCanvasTransformer::AddTransformer(CanvasTransformer *transformer) {
+  transformer_list_.push_back(transformer);
+}
+
+void LinkedTransformerCanvasTransformer::AddTransformer(CanvasTransformerList transformer_list) {
```

Comment:
I'd name the parameter suff sufficiently differently that it is hard to mix up transformer_list and transformer_list_.

(just 'list' is probably a sufficiently good name).